swarm/pss/notify/notify.go

```diff
@@ -172,11 +216,11 @@ func (c *Controller) Notify(name string, data []byte) error {
 }
 
 // adds an client address to the corresponding address bin in the notifier service
-// this method is not concurrency safe
+// this method is not concurrency safe, and will panic if called with a non-existing notification service name
```

Comment:
whenever we call `addToNotifier` and `remoteFromNotifier`, the notifier is known, so we would not need to look it up again.
In fact, if notifier.bins is a map, then we can get rid of these functions and just do
`notifier.bins[addr.Hex()] = sendBin` 
and `delete(notifier.bins, add.Hex()`

Reply:
Since this is a temporary solution for routing anyway I'll just go ahead and make this change.

But "let's just make it a map" has performance repercussions, and I believe we should always think twice about throwing cycles at code which might end up being executed a lot.

A quick test shows that there is a BIG difference between even sequential retrieval in maps versus arrays. Retrieval might happen often, add to and remove from notifier less so.

Have a look:

https://gist.github.com/nolash/07907e493e06060251b71ea493f36522

```
$ go test -v -cpu 4 ./maparraybench/ -run ^$ -bench .*
goos: linux
goarch: amd64
BenchmarkMap-4     	    1000	   1821025 ns/op
BenchmarkArray-4   	   50000	     31315 ns/op
PASS
ok  	_/home/lash/tmp/maparraybench	4.059s
[lash@sostenuto tmp]$ vi maparraybench/maparraybench_test.go 
[lash@sostenuto tmp]$ vi maparraybench/maparraybench_test.go 
[lash@sostenuto tmp]$ go test -v -cpu 4 ./maparraybench/ -run ^$ -bench .*
benchmark with 100000 items
goos: linux
goarch: amd64
BenchmarkMap-4     	    1000	   1796776 ns/op
BenchmarkArray-4   	   50000	     31180 ns/op
PASS
ok  	_/home/lash/tmp/maparraybench	4.049s
$ vi maparraybench/maparraybench_test.go 
[lash@sostenuto tmp]$ go test -v -cpu 4 ./maparraybench/ -run ^$ -bench .*
benchmark with 100 items
goos: linux
goarch: amd64
BenchmarkMap-4     	 1000000	      1677 ns/op
BenchmarkArray-4   	30000000	        47.2 ns/op
PASS
ok  	_/home/lash/tmp/maparraybench	3.175s
```